### DISCUSSION TITLE
query parameter as Optional<String> within declarative HttpClient

### DISCUSSION THREAD

**Discussion by krasv** [micronaut-projects/micronaut-core#4849] [Q&A] [answered]
Hi, is there a limitation _not_ using java Optional as HttpClient parameter?

having the client declared as
```
@Client("/path")
public interface MyClient {
  @Get("/data")
  @Produces({"application/json"})
  HttpResponse<Data> getData(
      @Nonnull
      @QueryValue(value="param")
      Optional<String> param
  );
}
```

and invoking the client as such

```
   Data data = client.getData(Optional.of("paramValue"));
```

produces at the outbound request the string representation of the `Optional` instance instead of the optional value.

`GET /path/data?param=Optional[paramValue]`

The conversion is fixed set to String within the DefaultBinderRegistry see 
https://github.com/micronaut-projects/micronaut-core/blob/b3eaf537b6bd1c87edc26403a204c45f673e1220/http-client-core/src/main/java/io/micronaut/http/client/bind/DefaultHttpClientBinderRegistry.java#L81

Is there either a reason __not__ supporting `Optional<String>` or is there a way to add a custom binder like https://docs.micronaut.io/latest/guide/index.html#customArgumentBinding for such common data types?

Thanks in advance,
Sven

**Comment by graemerocher** [comment, marked as answer]
Seems to be the case, although if you speak to the designers of the `Optional` type they say it should only really be used in return types

**Reply by krasv** [reply]
Thanks, got another response into that direction too: https://rules.sonarsource.com/java/tag/clumsy/RSPEC-3553

**Reply by krasv** [reply]
Just got another hint: version 2.2.2 addresses exact that issue: https://github.com/micronaut-projects/micronaut-core/issues/4568
